Update documuent viewer in the attachment widget on attribute change

signedav opened the pull request:
In the Attachment-Widget the document viewer type can be controlled over conditions. 

On update of an attribute in the form, the result of this condition is reevaluated and the document viewer type is updated on runtime.

If you set this expression:
```
CASE WHEN  "infofile"  ILIKE '%.jpg' AND "show_preview" 
THEN 
  'Image'
ELSE 
  CASE WHEN  "infofile"  ILIKE 'http%' AND "show_preview" 
  THEN 
    'Web'
  ELSE 
    'NoContent'
  END
END
```
Then you will get this behavior:
![document_viewer](https://user-images.githubusercontent.com/28384354/90361235-f37f2d80-e05d-11ea-81ca-6c3677e0e2ca.gif)

On change of an attribute the documentViewer expression of the externalRessourceWidget depends on, the documentViewer should be updated.

To not be needed to check for all the widget values in the form, the mCurrentFormFeature is introduced that contains all the values that are received on change of the widget values.

Resolves #37678

signedav committed:
On change of an attribute the documentViewer expression of the externalRessourceWidget depends on, the documentViewer should be updated.
To not be needed to check for all the widget values in the form, the mCurrentFormFeature is introduced that contains all the values that are received on change of the widget values.

Resolves #37678

signedav committed:
updateProperties function private and no reference for currentFormFeature

m-kuhn commented on python/gui/auto_generated/qgsattributeform.sip.in:
```diff
@@ -44,6 +44,14 @@ class QgsAttributeForm : QWidget
 
     const QgsFeature &feature();
 
+    const QgsFeature &currentFormFeature();
```
```suggestion
    QgsFeature currentFormFeature() const;
```

Better safe than sorry, returning refs has potential risks if used unconsciously and returning an implicitly shared construct like a feature is almost free.

signedav replied:
Okay.

m-kuhn commented:
Thanks @signedav 

@elpaso what do you think about keeping a "currentFormFeature" up to date this way?
I think we already have code to do the same in various places (e.g. for child forms, for drill down cascading, maybe more). Having it once in a centralized way sounds like the right thing to do?

m-kuhn commented on src/gui/qgsexternalresourcewidget.h:
```diff
@@ -143,6 +143,11 @@ class GUI_EXPORT QgsExternalResourceWidget : public QWidget
     void setDefaultRoot( const QString &defaultRoot );
 
+    /**
+     * Reloads the document from the path of the filewidget
```
since 3.16 ?

m-kuhn commented on src/gui/qgsattributeform.h:
```diff
@@ -75,6 +75,14 @@ class GUI_EXPORT QgsAttributeForm : public QWidget
+     *
+     * \since QGIS 3.16
+     */
+    const QgsFeature &currentFormFeature() { return mCurrentFormFeature; }
```
also needs updating here and in the header

signedav commented on src/gui/editorwidgets/qgsexternalresourcewidgetwrapper.cpp:
```diff
@@ -112,15 +112,25 @@ void QgsExternalResourceWidgetWrapper::setFeature( const QgsFeature &feature )
           dvc = QgsExternalResourceWidget::NoContent;
         }
         mQgsWidget->setDocumentViewerContent( dvc );
+        mQgsWidget->reloadDocument();
```
Hm... Yes, we could do that. Means I would remove `reloadDocument()` completely and call `loadDocument(path)` in the `setDocumentViewerContent`.

m-kuhn replied:
Less risk of forgetting it after calling `setDocumentViewerContent`?

signedav replied:
Yes.

signedav committed:
call loadDocument in setDocumentViewerContent instead over a seperate method
updateDocumentViewer needs not to be called on Image before the laod of the document (because it's called after the document load (or unload))
